hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+      MODIFY, REMOVE
+    }
+
+    interface Operation {
+      void apply(FileSystem fs, Path path, List<AclEntry> aclList) throws IOException;
+    }
+
+    private FileSystem fs;
+    private Path path;
+    private List<AclEntry> dirAcl;
+    private List<AclEntry> fileAcl;
+    private boolean recursive;
+    private Operation operation;
+    private List<Path> skipPaths;
+
+    HDFSAclOperation(FileSystem fs, Path path, Set<String> users, OperationType operationType,
+        FsAction fsAction, boolean recursive, List<Path> skipPaths) {
+      this.fs = fs;
+      this.path = path;
+      this.dirAcl = getDefaultAclEntries(users, fsAction);
+      this.fileAcl = getAccessAclEntries(users, fsAction);
+      this.recursive = recursive;
+      this.skipPaths = skipPaths;
+      if (operationType == OperationType.MODIFY) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.modifyAclEntries(modifyPath, aclList);
+      } else if (operationType == OperationType.REMOVE) {
+        operation =
+            (fileSystem, modifyPath, aclList) -> fileSystem.removeAclEntries(modifyPath, aclList);
+      }
+    }
+
+    HDFSAclOperation(Path path, HDFSAclOperation parent) {
+      this.fs = parent.fs;
+      this.path = path;
+      this.dirAcl = parent.dirAcl;
+      this.fileAcl = parent.fileAcl;
+      this.operation = parent.operation;
+      this.recursive = parent.recursive;
+      this.skipPaths = parent.skipPaths;
+    }
+
+    List<HDFSAclOperation> getChildAclOperations() throws IOException {
+      List<HDFSAclOperation> hdfsAclOperations = new ArrayList<>();
+      if (!skipPaths.contains(path) && recursive && fs.isDirectory(path)) {
+        FileStatus[] fileStatuses = fs.listStatus(path);
+        for (FileStatus fileStatus : fileStatuses) {
+          hdfsAclOperations.add(new HDFSAclOperation(fileStatus.getPath(), this));
+        }
+      }
+      return hdfsAclOperations;
+    }
+
+    void setAcl() throws IOException {
+      if (!skipPaths.contains(path) && fs.exists(path)) {
+        if (fs.isDirectory(path)) {
+          if (recursive) {
+            operation.apply(fs, path, dirAcl);
+          } else {
+            operation.apply(fs, path, fileAcl);
+          }
+        } else {
+          operation.apply(fs, path, fileAcl);
+        }
+      }
+    }
+
+    private List<AclEntry> getAccessAclEntries(Set<String> users, FsAction action) {
+      List<AclEntry> aclList = new ArrayList<>();
+      for (String user : users) {
+        aclList.add(aclEntry(ACCESS, user, action));
+      }
+      return aclList;
+    }
+
+    private List<AclEntry> getDefaultAclEntries(Set<String> users, FsAction action) {
+      List<AclEntry> dirAclList = new ArrayList<>();
+      for (String user : users) {
+        dirAclList.add(aclEntry(ACCESS, user, action));
+        dirAclList.add(aclEntry(DEFAULT, user, action));
+      }
+      return dirAclList;
+    }
+  }
+
+  protected static final class PathHelper {
```

Comment:
Can be private ?

Reply:
Also used in SnapshotScannerHDFSAclController